README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
Quick start
-----------

Welcome to the `graflex` GitHub page!

The Fisher's Exact Test is often used to measure the significance of overlap between two categorical variables. However, the test operates on the assumption that the relationship between observations is independent. Depending on how a graph is made, this assumption may not hold, thus making the Fisher's Exact Test invalid. This package offers a fast way to permute a false discovery rate (FDR) for the edge overlap between two non-random graphs using an odds ratio (OR) test statistic.

``` r
library(devtools)
devtools::install_github("tpq/graflex")
```

``` r
library(graflex)
set.seed(1)
A <- randGraph(100) # a graph of interest
K <- randK(100, 3) # a database relating nodes (rows) to concepts (cols)
graflex(A, K) # permute FDR for overlap
#> |------------(25%)----------(50%)----------(75%)----------|
#>   Neither G.only A.only Both      Odds       LogOR FDR.under FDR.over
#> 1    1887    611   1838  614 1.0317003  0.03120821     0.664    0.354
#> 2    1870    628   1855  597 0.9583240 -0.04256932     0.260    0.762
#> 3    1864    634   1861  591 0.9336794 -0.06862220     0.166    0.844
#>   Permutes
#> 1      500
#> 2      500
#> 3      500
```


NEWS.md
## graflex 0.0.3
---------------------
* Fix bug where `getK` triggers an error

## graflex 0.0.2
---------------------
* Add tests for `binTab` and `getFDR` functions

## graflex 0.0.1
---------------------
* New backend functions
    * `sym` function makes matrix symmetric
    * `binTab` function provides fast alternative to `table`
    * `getOR` wraps `binTab` to calculate odds ratio
    * `packageCheck` and `progress` not exported
* New random functions
    * `randPois` function makes a random Poisson matrix
    * `randGraph` function makes a random adjacency matrix
    * `randConcept` function makes a random fully connected sub-graph
    * `randPropd` function makes an adjacency matrix from a `propr::propd` analysis
    * `randK` function makes a random knowledge database
* New OR functions
    * `calculateOR` calculates the actual odds ratio (OR)
    * `permuteOR` permutes many odds ratios (OR)
    * `getFDR` calculates FDR based on ORs
* New graflex functions
    * `getK` builds knowledge database for human genes
    * `graflex` calculates FDR for multiple concepts
* New vignette


tests/testthat.R
library(testthat)
library(graflex)

test_check("graflex")


tests/testthat/test-binTab.R
library(graflex)

A <- sample(0:1, size = 500, replace = TRUE)
B <- sample(0:1, size = 500, replace = TRUE)

test_that("binTab and table agree", {

  expect_equal(
    matrix(table(A, B), 2, 2),
    binTab(A, B)
  )
})


tests/testthat/test-getFDR.R
library(graflex)

set.seed(1)
A <- randGraph(50)
B <- randConcept(50)

test_that("", {

  a <- calculateOR(A, B)
  o <- permuteOR(A, B)
  fdr <- getFDR(a, o)

  expect_gte(
    fdr$FDR.under,
    .05
  )

  expect_gte(
    fdr$FDR.over,
    .05
  )
})


R/2-permute.R
#' Permute Odds Ratio
#'
#' This function permutes \code{p} odds ratios for the
#'  edge overlap between two non-random graphs.
#'  It does this by randomly shuffling the rows and
#'  columns of \code{A} (jointly, thus preserving
#'  the degree distribution).
#'
#' Note that this function calculates overlap for the
#'  lower-left triangle of the input matrices.
#'
#' @param A,G An adjacency matrix.
#' @param p An integer. The number of overlaps to permute.
#' @export
permuteOR <- function(A, G, p = 500){

  Gstar <- G[lower.tri(G)]
  res <- lapply(1:p, function(i){

    # Shuffle the adjacency matrix
    index <- sample(1:ncol(A))
    A <- A[index, index]
    Astar <- A[lower.tri(A)]
    getOR(Astar, Gstar)
  })

  do.call("rbind", res)
}

#' Calculate Odds Ratio
#'
#' This function calculates an odds ratio for the
#'  edge overlap between two non-random graphs.
#'
#' Note that this function calculates overlap for the
#'  lower-left triangle of the input matrices.
#'
#' @inheritParams permuteOR
#' @export
calculateOR <- function(A, G){

  Astar <- A[lower.tri(A)]
  Gstar <- G[lower.tri(G)]
  getOR(Astar, Gstar)
}

#' Calculate Odds Ratio FDR
#'
#' This function calculates the false discovery rate (FDR)
#'  for over- and under-enrichment by counting the number of
#'  times the actual OR was greater than
#'  (or less than) a permuted OR.
#'  Used by \code{\link{graflex}}.
#'
#' @param actual A result from \code{\link{calculateOR}}.
#' @param permuted A result from \code{\link{permuteOR}}.
#' @return A \code{data.frame} of the FDRs for over-
#'  and under- enrichment.
#' @export
getFDR <- function(actual, permuted){

  actual$FDR.under <- sum(permuted$Odds <= actual$Odds) / nrow(permuted)
  actual$FDR.over <- sum(permuted$Odds >= actual$Odds) / nrow(permuted)
  actual
}


R/3-graflex.R
#' Build Knowledge Database for Human Genes
#'
#' This helper function wraps the few steps needed
#'  to make a "knowledge database" used by \code{\link{graflex}}.
#'
#' A knowledge database contains a row (g) for each node in the graph,
#'  and a column (c) for each concept tested for enrichment.
#'  \code{K_{gc} = 1} whenever a node associates with a concept.
#'
#' @param keys A character vector. The IDs for the genes to map.
#'  Argument passed to \code{AnnotationDbi::select}.
#' @param columns A string. The concepts to return (e.g., GO).
#'  Argument passed to \code{AnnotationDbi::select}.
#' @param keytype A string. The type of IDs being mapped.
#'  Argument passed to \code{AnnotationDbi::select}.
#' @param minK An integer. The minimum number of nodes to which
#'  a concept maps to be included in final table.
#'  Skip with \code{minK = 0}.
#' @return A knowledge database where each row is a graph node
#'  and each column is a concept.
#' @export
getK <- function(keys, columns = "GO", keytype = "ENSEMBL", minK = 10){

  if(length(columns) > 1) stop("Please provide a single column.")

  packageCheck("AnnotationDbi")
  packageCheck("org.Hs.eg.db")

  # Build K knowledge database from GO
  db <- org.Hs.eg.db::org.Hs.eg.db
  godf <- AnnotationDbi::select(db, keys = keys, columns = columns, keytype = keytype)
  gotab <- table(godf[,1:2]) # handles NAs!!
  gotab[gotab > 1] <- 1

  # Sort and filter K database
  gotab <- gotab[keys, ]
  if(!identical(rownames(gotab), keys)){
    stop("Uh oh! Unexpected mapping.")
  }
  gotab[,colSums(gotab) >= minK]
}

#' Permute FDR for Multiple Concepts
#'
#' This function calls \code{\link{permuteOR}} for each
#'  concept (i.e., column) in the database \code{K}.
#'  See \code{\link{getK}} for more information.
#'
#' For each concept, this function calculates the
#'  false discovery rate (FDR) by counting the number of
#'  times the actual OR was greater than
#'  (or less than) a permuted OR.
#'
#' @inheritParams permuteOR
#' @param A An adjacency matrix.
#' @param K A knowledge database where each row is a graph node
#'  and each column is a concept.
#' @export
graflex <- function(A, K, p = 500){

  if(nrow(A) != nrow(K)) stop("'A' and 'K' must have identical rows.")

  numTicks <- 0
  res <- lapply(1:ncol(K), function(k){

    numTicks <<- progress(k, ncol(K), numTicks)
    Gk <- K[,k] %*% t(K[,k])
    actual <- calculateOR(A, Gk)
    permuted <- permuteOR(A, Gk, p = p)
    actual <- getFDR(actual, permuted)
    actual$Permutes <- p
    actual$Concept <- colnames(K)[k]
    actual
  })

  do.call("rbind", res)
}


R/1-randFuncs.R
#' Make Random Graph
#'
#' This function makes a random graph.
#'
#' @param size An integer. The size of the graph.
#' @return An adjacency matrix.
#' @export
randGraph <- function(size){

  A <- matrix(sample(c(0, 1), size^2, replace = TRUE), size, size)
  diag(A) <- 0
  sym(A)
}

#' Make Random Concept Graph
#'
#' This function makes a random concept graph, defined here
#'  as a fully connected sub-graph.
#'
#' @inheritParams randGraph
#' @param fill An integer. The number of nodes to connect.
#' @return An adjacency matrix.
#' @export
randConcept <- function(size, fill = .5*size){

  if(fill > size) stop("Please provide 'fill' smallers than 'size'.")

  g <- vector("numeric", size)
  index <- sample(1:length(g), fill)
  g[index] <- 1
  g %*% t(g)
}

#' Make Random Matrix
#'
#' This function makes a random matrix by sampling
#'  rows from a Poisson distribution.
#'
#' @param nrows An integer. The number of rows in the matrix.
#' @param ncols An integer. The number of columns in the matrix.
#' @param lambda Argument passed to \code{rpois}.
#' @return A matrix.
#' @export
randPois <- function(nrows, ncols = nrows, lambda = 100){

  counts <- stats::rpois(nrows*ncols, lambda = lambda)
  matrix(counts, nrows, ncols)
}

#' Make Random Propd Graph
#'
#' This function builds a graph based on a \code{propr::propd}
#'  analysis of a random Poisson matrix.
#'
#' @inheritParams randGraph
#' @return An adjacency matrix.
#' @export
randPropd <- function(size){

  packageCheck("propr")

  dat <- randPois(size)
  pd <- suppressMessages(
    propr::propd(dat, sample(c("A", "B"), replace = TRUE, size = size))
  )
  pd <- suppressMessages(
    propr::updateF(pd)
  )

  cutoff <- propr::qtheta(pd, pval = .05)
  A <- propr::getAdj(pd, cutoff = cutoff)
  A
}

#' Make Random Knowledge Database
#'
#' This function builds a knowledge database with nodes
#'  as rows and concepts as columns.
#'
#' @param nnodes The number of nodes (rows).
#' @param nconcepts The number of concepts (columns).
#' @inheritParams randConcept
#' @return A knowledge database where each row is a graph node
#'  and each column is a concept.
#' @export
randK <- function(nnodes, nconcepts, fill = .5*nnodes){

  K <- matrix(0, nnodes, nconcepts)
  K <- apply(K, 2, function(concept){
    index <- sample(1:length(concept), fill)
    concept[index] <- 1
    concept
  })
  K
}


R/0-backend.R
#' Symmetrize Matrix
#'
#' This function makes a square matrix symmetric.
#'
#' @param A An adjacency matrix.
#' @return An adjacency matrix.
#' @export
sym <- function(A){

  if(nrow(A) != ncol(A)) stop("Please provide a square matrix.")

  fill <- A[lower.tri(A)]
  A <- t(A)
  A[lower.tri(A)] <- fill
  A
}

#' Tabulate Overlap
#'
#' This function tabulates the overlap between
#'  two vectors or two adjacency matrices.
#'  It is a faster version of \code{table} that
#'  only supports binary input.
#'
#' @param A,G A vector or adjacency matrix.
#' @return A table of overlap.
#' @export
binTab <- function(A, G){

  diff <- A != G
  only1 <- A[diff]
  b <- sum(only1)
  c <- length(only1) - b

  same <- !diff
  double1 <- A[same]
  a <- sum(double1)
  d <- length(double1) - a

  matrix(c(d, b, c, a), 2, 2)
}

#' Calculate Odds Ratio
#'
#' This function calculates the overlap between
#'  two vectors or two adjacency matrices.
#'  It returns the OR as well as other metrics.
#'
#' @inheritParams binTab
#' @return A \code{data.frame} of results.
#' @export
getOR <- function(A, G){

  tab <- binTab(A, G)
  or <- (tab[1,1] * tab[2,2]) / (tab[1,2] * tab[2,1])
  data.frame(
    "Neither" = tab[1,1],
    "G.only" = tab[1,2],
    "A.only" = tab[2,1],
    "Both" = tab[2,2],
    "Odds" = or,
    "LogOR" = log(or)
  )
}

#' Make Progress Bar
#'
#' @param i The current iteration.
#' @param k Total iterations.
#' @param numTicks The result of \code{progress}.
#' @return The next \code{numTicks} argument.
progress <- function(i, k, numTicks){

  if(i == 1) numTicks <- 0

  if(numTicks == 0) cat("|-")

  while(i > numTicks*(k/40)){

    cat("-")
    if(numTicks == 10) cat("(25%)")
    if(numTicks == 20) cat("(50%)")
    if(numTicks == 30) cat("(75%)")
    numTicks <- numTicks + 1
  }

  if(i == k) cat("-|\n")

  return(numTicks)
}

#' Package Check
#'
#' Checks whether the user has the required package installed.
#'  For back-end use only.
#'
#' @param package A character string. An R package.
packageCheck <- function(package){

  if(!requireNamespace(package, quietly = TRUE)){
    stop("Uh oh! This graflex method depends on ", package, "! ",
         "Try running: install.packages('", package, "')")
  }
}
